### PR TITLE
Dynamic routing mark 2

### PR DESCRIPTION
**Pull request by Timer**
This implements a proposed alternative implementation of dynamic routing.

### CODE REVIEW

File: packages/next-server/lib/router/router.ts [hunk 1]
```diff
@@ -225,6 +229,23 @@ export default class Router implements BaseRouter {
       const route = toRoute(pathname)
       const { shallow = false } = options
 
+      // detect dynamic routing
+      if (route.indexOf('/$') !== -1) {
+        const { re: routeRegex, groups } = getRouteRegex(route)
+        const routeMatch = routeRegex.exec(as)
+        if (!routeMatch) {
+          console.error('TODO: ...')
```
Comment (ijjk) [inline packages/next-server/lib/router/router.ts]:
```suggestion
          console.error('Your `as` value does not match your `href` value which is invalid')
```

**Review by ijjk — APPROVED** [review]

**Comment by jagreehal** [comment]
@Timer In Nuxt you can define a route which is a catch all e.g. https://nuxtjs.org/guide/routing#dynamic-routes

Is this going to be possible with this PR e.g. if I create /post/{id}/_.js then /post/{id}/foo/bar would resolve to the catch all page?

**Comment by Timer** [comment]
@jagreehal can you please explain a valid use case for catch all?

**Comment by jagreehal** [comment]
So we're building an application which works as CMS allowing the user to create pages. So the user can specify the data to be shown, the template to use and the URL. 

Using Nuxt I'm able to give the url to the backend API which resolves the data and then I can dynamically load the template required.

So rather then define every route such as  '/post/123/information/about-us' I want to be able to create catch all route.  Using Nuxt this would be something like /post/_postId/_.vue.

With the PR that's been merged I can only go one level deep e.g. /post/$postId/$page.js but not two or more levels deep without explicitly defining routes.

Does that make sense?

**Comment by Timer** [comment]
@jagreehal I think this would be better implemented as a router feature than a Next.js feature -- are you deploying to now? You can configure this behavior in your `routes` key within `now.json`.

**Comment by jagreehal** [comment]
@Timer Not using now. I still have hope it could be a Next feature as it still needs access to route params but needs to resolve to the 'nearest' catch all page. (⚠️ not great suggestion ahead) What about allowing the regex for the page to accept more than just one param e.g. '*.js'. I'll try to look into Nuxt implementation for better suggestions

**Comment by Timer** [comment]
We'll support `_.js` as a fallback route -- see upcoming PRs.

**Comment by jagreehal** [comment]
@Timer thanks you so much. I'd like to help in anyway I can testing, documentation, etc. Just let me know.

**Comment by Xiphe** [comment]
Love to see this change coming to next! Thanks for the affords.

When tinkering with it I noticed that it's incompatible with [zones](https://nextjs.org/docs#multi-zones). -> https://github.com/zeit/next.js/issues/7488

**Pull request merged**